addons/mass_mailing/tests/test_mailing_internals.py

```diff
@@ -498,6 +499,54 @@ def test_process_mailing_queue_without_html_body(self):
         self.assertFalse(mailing.body_html)
         self.assertEqual(mailing.mailing_model_name, 'res.partner')
 
+    @users('user_marketing')
+    @mute_logger('odoo.addons.mail.models.mail_mail')
+    def test_mailing_campaign_test_url_replacement(self):
+        """ Test the urls after sending mail process for mailing campaign tests (internal, tracked and external links) """
+
+        base_url = self.env['mailing.mailing'].get_base_url()
+
+        mailing = self.env['mailing.mailing'].create({
+            'name': 'Test',
+            'subject': 'Test',
+            'body_html': f"""
+                <p><a id=\"internal_link\" href=\"{base_url}\">Internal Test Link</a></p>
+                <p><a id=\"internal_tracked_link\" href=\"{base_url}/r/001\">Internal Test Tracked Link</a></p>
+                <p><a id=\"external_link\" href=\"https://www.reddit.com/r/Odoo\">External Test Link</a></p>
+            """,
+            'contact_list_ids': [Command.set(self.mailing_list_1.ids)],
+        })
+
+        # We don't want to create link trackers for tests
+        def patched_get_link_tracker_values(self):
+            self.ensure_one()
+            return {}
+
+        self.assertEqual(self.mailing_list_1.contact_ids.message_ids, self.env['mail.message'])
+        with patch("odoo.addons.mass_mailing.models.mailing.MassMailing._get_link_tracker_values",
+                   new=patched_get_link_tracker_values):
+            with self.mock_mail_gateway(mail_unlink_sent=True):
+                mailing.action_send_mail()
+
+        self.assertTrue(self._mails)
+        new_mail = self._mails[0]
+
+        expected_links = {
+            'internal_tracked_link': base_url + '/r/001',
+            'internal_link': base_url,
+            'external_link': 'https://www.reddit.com/r/Odoo',
+        }
+
+        for id, link in expected_links.items():
+            anchor_href = self._get_href_from_anchor_id(new_mail['body'], id)
+            if anchor_href.startswith(base_url):
+                if '/r/' in link:
+                    self.assertIn(link + '/m/', anchor_href, "Tracked url should contains the base tracked url and a /m/ parameter")
+                else:
+                    self.assertEqual(link, anchor_href, "The url should not change if not tracked")
+            else:
+                self.assertEqual(link, anchor_href, "An external link can't be modified")
```

Comment:
This whole part is a bit hard to read. Since you only have 3 values to check, I think you could do away with the loop and just use three `assertIn` calls.